presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractGroupCollectionAggregationState.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.array.IntBigArray;
+import com.facebook.presto.array.ShortBigArray;
+import com.facebook.presto.operator.aggregation.state.AbstractGroupedAccumulatorState;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import it.unimi.dsi.fastutil.longs.LongList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Verify.verify;
+/**
+ * state object that uses a single BlockBuilder for all groups.
+ */
+public abstract class AbstractGroupCollectionAggregationState<T>
+        extends AbstractGroupedAccumulatorState
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(AbstractGroupCollectionAggregationState.class).instanceSize();
+    private static final int MAX_NUM_BLOCKS = 30000;
+    private static final short NULL = -1;
+
+    private final ShortBigArray headBlockIndex;
+    private final IntBigArray headPosition;
+
+    private final ShortBigArray nextBlockIndex;
+    private final IntBigArray nextPosition;
+
+    private final ShortBigArray tailBlockIndex;
+    private final IntBigArray tailPosition;
+
+    private final List<PageBuilder> values;
+    private final LongList sumPositions;
```

Comment:
or `sumPositionCounts` and add a comment? Since it's 
```
   sumPositionCounts[n] = \sum_{i=0}^{n-1} values[i].getPositionCount()
```

I suggest do the rename in a follow-up commit given the size of this commit :)

Reply:
Given that this is moved from earlier code, I will address this feedback separately.